pkg/controller/tas/topology_ungater.go

```diff
@@ -0,0 +1,300 @@
+		Named(TASTopologyUngater).
+		For(&kueue.Workload{}).
+		Watches(&corev1.Pod{}, &podHandler).
+		WithOptions(controller.Options{NeedLeaderElection: ptr.To(false)}).
+		WithEventFilter(r).
+		Complete(core.WithLeadingManager(mgr, r, &kueue.ClusterQueue{}, cfg))
+}
+
+var _ handler.EventHandler = (*podHandler)(nil)
+
+type podHandler struct {
+}
+
+func (h *podHandler) Create(_ context.Context, e event.CreateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := e.Object.(*corev1.Pod)
+	if !isPod {
+		return
+	}
+	h.queueReconcileForPod(pod, q)
+}
+
+func (h *podHandler) Update(ctx context.Context, e event.UpdateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	oldPod, isOldPod := e.ObjectOld.(*corev1.Pod)
+	newPod, isNewPod := e.ObjectNew.(*corev1.Pod)
+	if !isOldPod || !isNewPod {
+		return
+	}
+	h.queueReconcileForPod(oldPod, q)
+	h.queueReconcileForPod(newPod, q)
+}
+
+func (h *podHandler) Delete(_ context.Context, e event.DeleteEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := e.Object.(*corev1.Pod)
+	if !isPod {
+		return
+	}
+	h.queueReconcileForPod(pod, q)
+}
+
+func (h *podHandler) queueReconcileForPod(pod *corev1.Pod, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	if pod == nil {
+		return
+	}
+	if !utilpod.HasGate(pod, kueuealpha.TopologySchedulingGate) {
+		return
+	}
+	if wlName, found := pod.Annotations[kueuealpha.WorkloadAnnotation]; found {
+		q.AddAfter(reconcile.Request{NamespacedName: types.NamespacedName{
+			Name:      wlName,
+			Namespace: pod.Namespace,
+		}}, ungateBatchPeriod)
+	}
+}
+
+func (h *podHandler) Generic(context.Context, event.GenericEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (r *topologyUngater) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("workload", req.NamespacedName.Name)
+	log.V(2).Info("Reconcile Topology Ungater")
+
+	wl := &kueue.Workload{}
+	if err := r.client.Get(ctx, req.NamespacedName, wl); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return reconcile.Result{}, err
+		}
+		log.Info("workload not found")
+		return reconcile.Result{}, nil
+	}
+	if wl.Status.Admission == nil {
+		log.Info("workload is not admitted")
+		return reconcile.Result{}, nil
+	}
+
+	allToUngate := make([]podWithUngateInfo, 0)
+	for _, psa := range wl.Status.Admission.PodSetAssignments {
+		if psa.TopologyAssignment != nil {
+			toUngate, err := r.podsetPodsToUngate(ctx, log, wl, &psa)
+			if err != nil {
+				log.Error(err, "failed to identify pods to ungate", "podset", psa.Name, "count", psa.Count)
+				return reconcile.Result{}, err
+			} else {
+				log.Info("identified pods to ungate for podset", "podset", psa.Name, "count", len(toUngate))
+				allToUngate = append(allToUngate, toUngate...)
+			}
+		}
+	}
+	var err error
+	if len(allToUngate) > 0 {
```

Comment:
How do you imagine the decoupling? One idea I have is to move out the code which for a list of pods ungates them all. It would be very similar to the code we have in pod_controller which for a list of pods cleans up finalizers: https://github.com/kubernetes-sigs/kueue/blob/4a5155ac1b39e82b26ff8e64bab09b31dc724995/pkg/controller/jobs/pod/pod_controller.go#L885-L919. In any case, I would prefer to move the refactoring to a follow up.

Reply:
I'm wondering if we can consolidate expectations and podWithUngateInfo into the TAS dedicated cache.
And then, we can decouple TASController including reconciler, eventHandler, and Predicator, and TASUngator including this ungating processes.
Becuase ideally, in general, we should not spawn the separate go routine in the reconciler. The separate Go routine could delay the next Reconcile and, the object information in reconcile queue could older.